File: app/src/main/java/com/jjf/template/repository/UserRepository.java
package com.jjf.template.repository;

import com.jjf.template.AppExecutors;
import com.jjf.template.api.ApiResponse;
import com.jjf.template.api.GithubService;
import com.jjf.template.bean.Resource;
import com.jjf.template.bean.User;
import com.jjf.template.db.UserDao;

import javax.inject.Inject;
import javax.inject.Singleton;

import androidx.lifecycle.LiveData;

/**
 * @author xj
 * date: 18-11-8
 * description :
 */
@Singleton
public class UserRepository {

    private final GithubService mService;
    private final AppExecutors mAppExecutors;
    private final UserDao userDao;

    @Inject
    public UserRepository(AppExecutors appExecutors,GithubService service,UserDao userDao) {
        this.mAppExecutors = appExecutors;
        this.mService = service;
        this.userDao = userDao;
    }

    public LiveData<Resource<User>> loadUser(String login){
       return new NetworkBoundResource<User, User>(mAppExecutors) {
           @Override
           public LiveData<User> loadFromDb() {
               return userDao.findByLogin(login);
           }

           @Override
           protected boolean shouldFetch(User data) {
               return data == null;
           }

           @Override
           protected void saveCallResult(User item) {
               userDao.insert(item);
           }

           @Override
           protected LiveData<ApiResponse<User>> createCall() {
               return mService.getUser(login);
           }
       }.asLiveData();
    }
}

File: app/src/main/java/com/jjf/template/repository/UserRepository.kt
package com.jjf.template.repository

import androidx.lifecycle.LiveData
import com.jjf.template.AppExecutors
import com.jjf.template.data.api.ApiResponse
import com.jjf.template.data.api.GithubService
import com.jjf.template.result.Resource
import com.jjf.template.result.User
import com.jjf.template.data.db.UserDao
import javax.inject.Inject
import javax.inject.Singleton

/**
 * @author xj
 * date: 18-11-8
 * description :
 */
@Singleton
class UserRepository @Inject
constructor(private val mAppExecutors: AppExecutors, private val mService: GithubService, private val userDao: UserDao) {

    fun loadUser(login: String): LiveData<Resource<User>> {
        return object : NetworkBoundResource<User, User>(mAppExecutors) {
            override fun shouldFetch(data: User?): Boolean {
                return data == null
            }

            override fun loadFromDb(): LiveData<User> {
                return userDao.findByLogin(login)
            }

            override fun saveCallResult(item: User) {
                userDao.insert(item)
            }

            override fun createCall(): LiveData<ApiResponse<User>> {
                return mService.getUser(login)
            }
        }.asLiveData()
    }
}
